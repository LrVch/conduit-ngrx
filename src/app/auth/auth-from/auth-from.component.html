<form [formGroup]="form"
      (ngSubmit)="submit()">

  <mat-form-field *ngIf="authType === 'register'"
                  class="input-full-width">
    <input matInput
           formControlName="username"
           placeholder="Username"
           autocomplete="none"
           type="text">
  </mat-form-field>

  <mat-form-field class="input-full-width">
    <input matInput
           formControlName="email"
           placeholder="Email"
           autocomplete="none"
           type="text">
  </mat-form-field>

  <mat-form-field class="input-full-width">
    <input matInput
           formControlName="password"
           placeholder="Password"
           autocomplete="none"
           type="password">
  </mat-form-field>
  <br>
  <button type="submit"
          [disabled]="_isSubmitting"
          mat-flat-button
          color="primary">title</button>
  <br>
  <br>

</form>