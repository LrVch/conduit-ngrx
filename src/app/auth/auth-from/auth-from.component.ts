import { Component, OnInit, Input, Output, EventEmitter } from '@angular/core';
import { FormGroup, FormBuilder, Validators, FormControl } from '@angular/forms';
import { Credentials } from 'src/app/core/models/credentials.model';

export interface AuthPayload {
  credentials: Credentials;
  authType: string;
}

@Component({
  selector: 'app-auth-from',
  templateUrl: './auth-from.component.html'
})

export class AuthFromComponent implements OnInit {
  @Input('authType') authType: string;
  @Output() submitForm = new EventEmitter<AuthPayload>();
  @Input() set isSubmitting(isSubmitting: boolean) {
    this._isSubmitting = isSubmitting;
    this.toggleDisableFields(this.form, isSubmitting);
  }

  _isSubmitting: boolean;

  form: FormGroup;

  constructor(
    private fb: FormBuilder,
  ) { }

  ngOnInit() {
    this.form = this.fb.group({
      'email': ['', Validators.required],
      'password': ['', Validators.required]
    });

    if (this.authType === 'register') {
      this.form.addControl('username', new FormControl('', Validators.required));
    }
  }

  submit() {
    if (this.form.valid) {
      const authPayload: AuthPayload = {
        credentials: this.form.value,
        authType: this.authType
      };

      this.submitForm.emit(authPayload);
    } else {
      this.validateAllFormFields(this.form);
    }
  }

  validateAllFormFields(formGroup: FormGroup): void {
    Object.keys(formGroup.controls).forEach(field => {
      const control = formGroup.get(field);
      if (control instanceof FormControl) {
        control.markAsTouched({ onlySelf: true });
      } else if (control instanceof FormGroup) {
        this.validateAllFormFields(control);
      }
    });
  }

  toggleDisableFields(formGroup: FormGroup, disable: boolean): void {
    if (!formGroup) {
      return;
    }
    Object.keys(formGroup.controls).forEach(field => {
      const control = formGroup.get(field);
      if (control instanceof FormControl) {
        if (disable) {
          control.disable();
        } else {
          control.enable();
        }
      } else if (control instanceof FormGroup) {
        this.validateAllFormFields(control);
      }
    });
  }
}
