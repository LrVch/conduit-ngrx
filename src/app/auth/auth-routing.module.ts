import { NgModule } from '@angular/core';
import { Routes, RouterModule } from '@angular/router';
import { AuthComponent } from './auth.component';
import { NoAuthGuard } from './no-auth-guard.service';

const routes: Routes = [
  {
    path: 'login',
    component: AuthComponent,
    canActivate: [NoAuthGuard],
    data: { title: 'conduit.menu.login', animation: 'AuthComponent' }
  },
  {
    path: 'register',
    component: AuthComponent,
    canActivate: [NoAuthGuard],
    data: { title: 'conduit.menu.register', animation: 'AuthComponent' }
  }
];

@NgModule({
  imports: [RouterModule.forChild(routes)],
  exports: [RouterModule]
})
export class AuthRoutingModule {}
