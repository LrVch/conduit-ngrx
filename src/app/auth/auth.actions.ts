import { Action } from '@ngrx/store';
import { Credentials } from '@app/core/models/credentials.model';
import { User, Errors } from '@app/core';

export enum AuthActionTypes {
  LoginPageAttemptLogin = '[Login Page] Attempt Login',
  LoginSuccess = '[Auth API] Login success',
  LoginFail = '[Auth API] Login fail',
  LoginAttemptCopmlete = '[Auth API] Login Attempt Copmlete',
  LoginPageSetAuthErrors = '[Auth API] Set Auth Errors',
  LoginPageClearAuthErrors = '[Login Page] Clear Auth Errors',
  LoggedLocalStorageRequest = '[LocalStorage] Login success Request',
  LoggedLocalStorage = '[LocalStorage] Login Success',
  ClearAuthErrors = '[Login Page] Clear Auth Errors',
  SettingsPageLogoutAction = '[Settings Page] Logout',
  LogoutConfirm = '[Auth] Logout Confirm',
  LogoutAction = '[Auth] Logout',
  AuthAttemptToGetUser = '[Auth] Try To Get User From Api',
  SetReturnUrl = '[Auth] Set Return Url',
  ClearReturnStateFromRouteChange = '[Auth] Clear Return State From Route Change',

  UpdateUserRequest = '[Auth API] Update User Request',
  UpdateUserSuccess = '[Auth API] Update User Success',
  UpdateUserFail = '[Auth API] Update User Fail',
  SetUpdateUserErrors = '[Auth API] Set Update User Errors'
}

export class SetReturnUrl implements Action {
  readonly type = AuthActionTypes.SetReturnUrl;

  constructor(public payload: { returnUrl: string }) { }
}

export class ClearReturnStateFromRouteChange implements Action {
  readonly type = AuthActionTypes.ClearReturnStateFromRouteChange;
}

export class AuthAttemptToGetUser implements Action {
  readonly type = AuthActionTypes.AuthAttemptToGetUser;
}

export class LoginSuccess implements Action {
  readonly type = AuthActionTypes.LoginSuccess;

  constructor(public payload: { user: User }) { }
}

export class LoggedLocalStorage implements Action {
  readonly type = AuthActionTypes.LoggedLocalStorage;

  constructor(public payload: { user: User }) { }
}

export class LoggedLocalStorageRequest implements Action {
  readonly type = AuthActionTypes.LoggedLocalStorageRequest;
}

export class LoginFail implements Action {
  readonly type = AuthActionTypes.LoginFail;

  constructor(public payload: { authErrors: Errors }) { }
}

export class LoginAttemptCopmlete implements Action {
  readonly type = AuthActionTypes.LoginAttemptCopmlete;
}

export class SettingsPageLogoutAction implements Action {
  readonly type = AuthActionTypes.SettingsPageLogoutAction;

  constructor(public payload: { question: string }) { }
}

export class LogoutConfirm implements Action {
  readonly type = AuthActionTypes.LogoutConfirm;
}

export class LogoutAction implements Action {
  readonly type = AuthActionTypes.LogoutAction;

  constructor(public payload?: { path?: string }) { }
}

export class LoginPageAttemptLogin implements Action {
  readonly type = AuthActionTypes.LoginPageAttemptLogin;

  constructor(public payload: { authType: string, credentials: Credentials }) { }
}

export class LoginPageClearAuthErrors implements Action {
  readonly type = AuthActionTypes.LoginPageClearAuthErrors;
}

export class ClearAuthErrors implements Action {
  readonly type = AuthActionTypes.ClearAuthErrors;
}


export class LoginPageSetAuthErrors implements Action {
  readonly type = AuthActionTypes.LoginPageSetAuthErrors;

  constructor(public payload: { authErrors: Errors }) { }
}

export class SetUpdateUserErrors implements Action {
  readonly type = AuthActionTypes.SetUpdateUserErrors;

  constructor(public payload: { authErrors: Errors }) { }
}

export class UpdateUserRequest implements Action {
  readonly type = AuthActionTypes.UpdateUserRequest;

  constructor(public payload: { user: User }) { }
}

export class UpdateUserSuccess implements Action {
  readonly type = AuthActionTypes.UpdateUserSuccess;

  constructor(public payload: { user: User }) { }
}

export class UpdateUserFail implements Action {
  readonly type = AuthActionTypes.UpdateUserFail;

  constructor(public payload: { errors: Errors }) { }
}

export type AuthActions =
  SetReturnUrl
  | ClearReturnStateFromRouteChange
  | AuthAttemptToGetUser
  | LoginSuccess
  | LoginFail
  | SettingsPageLogoutAction
  | LogoutConfirm
  | LogoutAction
  | LoginAttemptCopmlete
  | LoginPageAttemptLogin
  | LoggedLocalStorage
  | LoggedLocalStorageRequest
  | LoginPageClearAuthErrors
  | ClearAuthErrors
  | LoginPageSetAuthErrors
  | UpdateUserRequest
  | UpdateUserSuccess
  | UpdateUserFail
  | SetUpdateUserErrors;


