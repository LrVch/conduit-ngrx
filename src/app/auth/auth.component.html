<div class="auth-page page">
  <div class="container">
    <div class="row">

      <div class="col-md-6 offset-md-3 col-xs-12">
        <h1>{{ title }}</h1>
        <p>
          <a [routerLink]="['/login']"
             *ngIf="authType === 'register'">Have an account?</a>
          <a [routerLink]="['/register']"
             *ngIf="authType === 'login'">Need an account?</a>
        </p>

        <app-list-error [errors]="authErrors$ | async"></app-list-error>

        <form [formGroup]="authForm"
              (ngSubmit)="submitForm()">
          <fieldset [disabled]="isSubmitting$ | async">
            <fieldset class="">
              <mat-form-field *ngIf="authType === 'register'"
                              class="input-full-width">
                <input matInput
                       formControlName="username"
                       placeholder="Username"
                       type="text">
              </mat-form-field>
            </fieldset>
            <fieldset>
              <mat-form-field class="input-full-width">
                <input matInput
                       formControlName="email"
                       placeholder="Email"
                       type="text">
              </mat-form-field>
            </fieldset>
            <fieldset>
              <mat-form-field class="input-full-width">
                <input matInput
                       formControlName="password"
                       placeholder="Password"
                       type="password">
              </mat-form-field>
            </fieldset>
            <br>
            <button type="submit" [disabled]="!authForm.valid || (isSubmitting$ | async)" mat-flat-button color="primary"> {{ title }}</button>
            <br>
            <br>
          </fieldset>
        </form>
      </div>

    </div>
  </div>
</div>