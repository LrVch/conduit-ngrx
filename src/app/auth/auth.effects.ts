import { Injectable } from '@angular/core';
import { Actions, Effect, ofType } from '@ngrx/effects';
import {
  LoginSuccess,
  AuthActionTypes,
  LogoutAction,
  LoginFail,
  LoggedLocalStorage,
  LoginPageAttemptLogin,
  LoginAttemptCopmlete,
  LoginPageSetAuthErrors,
  UpdateUserRequest,
  UpdateUserSuccess,
  UpdateUserFail,
  SetUpdateUserErrors,
  LoggedLocalStorageRequest
} from './auth.actions';
import { tap, finalize, map, switchMap, catchError } from 'rxjs/operators';
import { Router } from '@angular/router';
import { Store, Action } from '@ngrx/store';
import { of, noop, defer, Observable } from 'rxjs';
import { UserService, User, Errors, ErrorsObj } from '../core';
import { AppState } from '../reducers';
import { JwtService } from '../core';
import { HideMainLoader } from '../layout/layout.actions';


@Injectable()
export class AuthEffects {

  @Effect({ dispatch: false })
  login$ = this.actions$.pipe(
    ofType<LoginSuccess>(AuthActionTypes.LoginSuccess),
    tap(action => {
      const { user } = action.payload;

      this.jwtService.saveToken(user.token);
      this.router.navigateByUrl('/');
    })
  );

  @Effect({ dispatch: false })
  logout$ = this.actions$.pipe(
    ofType<LogoutAction>(
      AuthActionTypes.LogoutAction,
    ),
    tap(() => {
      this.jwtService.destroyUseData();
      this.router.navigateByUrl('/login');
    })
  );

  @Effect({ dispatch: false })
  loginFail$ = this.actions$.pipe(
    ofType<LoginFail>(
      AuthActionTypes.LoginFail
    ),
    tap((action) => {
      const { authErrors } = action.payload;
      // console.error(authErrors);
      this.store.dispatch(new LoginPageSetAuthErrors({ authErrors }));
      this.store.dispatch(new HideMainLoader());
      this.jwtService.destroyUseData();
    })
  );

  @Effect()
  init$: Observable<Action> = defer((): Observable<LoggedLocalStorage | LoginFail> => {
    const token = this.jwtService.getToken();

    if (token) {
      this.store.dispatch(new LoggedLocalStorageRequest());
      return this.userService.getUser()
        .pipe(
          switchMap(responce => {
            const { user } = responce;
            return of(new LoggedLocalStorage({ user }));
          }),
          catchError(authErrors => {
            return of(new LoginFail({ authErrors }));
          })
        );
    } else {
      return of(new LoginFail({ authErrors: new ErrorsObj({ type: 'token', body: ['token not found'] }) }));
    }
  });

  @Effect()
  loginAttempt = this.actions$.pipe(
    ofType<LoginPageAttemptLogin>(AuthActionTypes.LoginPageAttemptLogin),
    switchMap(action => {
      return this.userService.attemptAuth(action.payload.authType, action.payload.credentials)
        .pipe(
          map(user => new LoginSuccess({ user })),
          catchError(authErrors => {
            return of(new LoginFail({ authErrors }));
          })
        );
    })
  );

  @Effect()
  updateInfo = this.actions$.pipe(
    ofType<UpdateUserRequest>(AuthActionTypes.UpdateUserRequest),
    switchMap(action => {
      return this.userService.update(action.payload.user)
        .pipe(
          map(user => new UpdateUserSuccess({ user })),
          catchError((errors: Errors) => {
            return of(new UpdateUserFail({ errors }));
          })
        );
    })
  );

  @Effect()
  updateInfoFail: Observable<Action> = this.actions$.pipe(
    ofType<UpdateUserFail>(AuthActionTypes.UpdateUserFail),
    map(action => new SetUpdateUserErrors({ authErrors: action.payload.errors }))
  );

  constructor(
    private actions$: Actions,
    private router: Router,
    private userService: UserService,
    private store: Store<AppState>,
    private jwtService: JwtService
  ) { }
}
