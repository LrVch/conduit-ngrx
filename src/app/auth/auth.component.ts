import { Component, OnInit, } from '@angular/core';
import { ActivatedRoute } from '@angular/router';

import { Errors } from '../core';
import { Store, select } from '@ngrx/store';
import { AppState } from '../reducers';
import { LoginPageAttemptLogin, LoginPageClearAuthErrors } from './auth.actions';
import { Observable } from 'rxjs';
import { selectAuthLoading, selectAuthErrors } from './auth.selectors';
import { ShowMainLoader } from '../layout/layout.actions';
import { AuthPayload } from './auth-from/auth-from.component';
import { map, tap } from 'rxjs/operators';

@Component({
  templateUrl: './auth.component.html'
})
export class AuthComponent implements OnInit {
  authType$: Observable<string>;
  title$: Observable<string>;
  authErrors$: Observable<Errors>;
  isSubmitting$: Observable<boolean>;

  constructor(
    private route: ActivatedRoute,
    private store: Store<AppState>
  ) { }

  ngOnInit() {
    this.authType$ = this.route.url.pipe(
      map(data => data[data.length - 1].path)
    );

    this.title$ = this.authType$.pipe(
      map(authType => authType === 'login' ? 'Sign in' : 'Sign up')
    );

    this.isSubmitting$ = this.store.pipe(
      select(selectAuthLoading)
    );

    this.authErrors$ = this.store.pipe(
      select(selectAuthErrors)
    );

    this.store.dispatch(new LoginPageClearAuthErrors());
  }

  submitForm(authPayload: AuthPayload): void {
    const { credentials, authType } = authPayload;

    this.store.dispatch(new LoginPageAttemptLogin({
      authType: authType,
      credentials
    }));

    this.store.dispatch(new ShowMainLoader());
  }
}
