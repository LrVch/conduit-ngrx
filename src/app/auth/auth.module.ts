import { NgModule } from '@angular/core';

import { AuthComponent } from './auth.component';
import { NoAuthGuard } from './no-auth-guard.service';
import { SharedModule } from '../shared';
import { AuthRoutingModule } from './auth-routing.module';
import { StoreModule } from '@ngrx/store';
import * as fromAuth from './auth.reducer';
import { EffectsModule } from '@ngrx/effects';
import { AuthEffects } from './auth.effects';

@NgModule({
  imports: [
    SharedModule,
    AuthRoutingModule,
    StoreModule.forFeature('auth', fromAuth.authReducer),
    EffectsModule.forFeature([AuthEffects])
  ],
  declarations: [
    AuthComponent
  ],
  providers: [
    NoAuthGuard
  ]
})
export class AuthModule {}
