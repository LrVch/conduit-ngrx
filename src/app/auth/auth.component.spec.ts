import { async, ComponentFixture } from '@angular/core/testing';
import { ConfigureFn, configureTests, RouterLinkDirectiveStubDirective, getAuthErrors } from '../lib/testing';
import { AuthComponent } from './auth.component';
import { DebugElement, Component, Output, EventEmitter, Input } from '@angular/core';
import { Errors } from 'src/app/core';
import { NoopAnimationsModule } from '@angular/platform-browser/animations';
import { By } from '@angular/platform-browser';
import { AuthPayload } from './auth-from/auth-from.component';
import { ActivatedRoute } from '@angular/router';
import { BehaviorSubject } from 'rxjs';
import { Store, StoreModule, combineReducers } from '@ngrx/store';
import * as Actions from './auth.actions';
import * as fromRoot from '../reducers';
import * as fromAuth from './auth.reducer';
import { RouterTestingModule } from '@angular/router/testing';
import { ShowMainLoader } from '../layout/layout.actions';
import { hot, cold } from 'jasmine-marbles';
import { first } from 'rxjs/operators';

const payload: AuthPayload = {
  credentials: {
    email: 'email',
    password: 'password'
  },
  authType: 'login'
};

@Component({
  selector: 'app-auth-from',
  template: `<div (click)="submit()">a</div>`
})
class CommentFormComponent {
  @Input('authType') authType: string;
  @Output() submitForm = new EventEmitter<AuthPayload>();
  @Input('isSubmitting') isSubmitting: boolean;

  submit(): void {
    const authPayload: AuthPayload = payload;

    this.submitForm.emit(authPayload);
  }
}

@Component({
  selector: 'app-list-error',
  template: ``
})
class ListErrorComponent {
  @Input('errors') errors: Errors;
}

describe('AuthComponent', () => {
  let component: AuthComponent;
  let fixture: ComponentFixture<AuthComponent>;
  let de: DebugElement;
  let store: Store<fromAuth.AuthState>;
  let route: ActivatedRoute;
  let linkDes;
  let routerLinks;

  class UrlSegment {
    constructor(public path: string) {
    }
  }

  class MockRoute {
    setPath = new BehaviorSubject([new UrlSegment('login')]);
    url = this.setPath.asObservable();
  }

  beforeEach(
    async(() => {
      const configure: ConfigureFn = testBed => {
        testBed.configureTestingModule({
          declarations: [AuthComponent, CommentFormComponent, ListErrorComponent, RouterLinkDirectiveStubDirective],
          imports: [
            NoopAnimationsModule,
            StoreModule.forRoot({
              ...fromRoot.reducers,
              feature: combineReducers(fromAuth.authReducer),
            }),
            RouterTestingModule
          ],
          providers: [
            {
              provide: ActivatedRoute, useClass: MockRoute
            }
          ]
        });
      };

      configureTests(configure).then(testBed => {
        fixture = testBed.createComponent(AuthComponent);
        component = fixture.componentInstance;
        de = fixture.debugElement;
        store = testBed.get(Store);
        route = testBed.get(ActivatedRoute);
        fixture.detectChanges();

        linkDes = de.queryAll(By.directive(RouterLinkDirectiveStubDirective));
        routerLinks = linkDes.map(deel => deel.injector.get(RouterLinkDirectiveStubDirective));

        spyOn(store, 'dispatch').and.callThrough();
      });
    })
  );

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should compile and show "Need an account?" link', () => {
    component.authType$.subscribe(res => {
      (<any>expect(fixture)).toMatchSnapshot();
    });
  });

  it('should compile and show "Have an account?" link', () => {
    (<any>route).setPath.next([new UrlSegment('register')]);

    fixture.detectChanges();

    component.authType$.subscribe(res => {
      (<any>expect(fixture)).toMatchSnapshot();
    });
  });

  it('can get RouterLinks from template for login route', () => {
    fixture.detectChanges();

    expect(routerLinks.length).toBe(1, 'should have 1 routerLink');
    expect(routerLinks[0].linkParams).toEqual(['/register']);
  });

  it('can get RouterLinks from template for register route', done => {
    (<any>route).setPath.next([new UrlSegment('register')]);
    fixture.detectChanges();

    linkDes = de.queryAll(By.directive(RouterLinkDirectiveStubDirective));
    routerLinks = linkDes.map(deel => deel.injector.get(RouterLinkDirectiveStubDirective));

    component.authType$.subscribe(res => {
      expect(routerLinks.length).toBe(1, 'should have 1 routerLink');
      expect(routerLinks[0].linkParams).toEqual(['/login']);
      done();
    }, done, done);
  });

  it('should dispatch "LoginPageAttemptLogin" and "ShowMainLoader"', () => {
    const loginPageAttemptLogin = new Actions.LoginPageAttemptLogin(payload);
    const showMainLoader = new ShowMainLoader();

    const submitDe = de.query(By.css('app-auth-from')).query(By.css('div'));

    submitDe.triggerEventHandler('click', null);

    expect(store.dispatch).toHaveBeenCalledWith(loginPageAttemptLogin);
    expect(store.dispatch).toHaveBeenCalledWith(showMainLoader);
  });

  it('should dispatch a "LoginPageClearAuthErrors" to clear auth errors when created', () => {
    const action = new Actions.LoginPageClearAuthErrors();
    component.ngOnInit();

    expect(store.dispatch).toHaveBeenCalledWith(action);
  });

  it('should get current route from Route', () => {
    component.authType$.pipe(first()).subscribe(res => {
      expect(res).toBe('login');
    });

    (<any>route).setPath.next([new UrlSegment('register')]);

    component.authType$.subscribe(res => {
      expect(res).toBe('register');
    });
  });

  it('should get title', () => {
    component.title$.pipe(first()).subscribe(res => {
      expect(res).toBe('Sign in');
    });

    (<any>route).setPath.next([new UrlSegment('register')]);

    component.title$.subscribe(res => {
      expect(res).toBe('Sign up');
    });
  });

  // Another was to test selected data from store
  // https://github.com/ngrx/platform/blob/master/docs/store/testing.md
  // it('should select isSubmitting', () => {
  //   const action = new Actions.LoginPageAttemptLogin(payload);

  //   store.dispatch(action);

  //   component.isSubmitting$.subscribe(res => {
  //     expect(res).toBeTruthy();
  //   });
  // });

  it('should select isSubmitting', () => {
    const spy = spyOn(store, 'pipe');

    spy.and.returnValue(hot('-a', { a: true }));

    component.ngOnInit();

    const expected = cold('-a', { a: true });

    expect(component.isSubmitting$).toBeObservable(expected);
  });

  it('should select authErrors', () => {
    const spy = spyOn(store, 'pipe');
    const errors = getAuthErrors();

    spy.and.returnValue(hot('-a', { a: errors }));

    component.ngOnInit();

    const expected = cold('-a', { a: errors });

    expect(component.authErrors$).toBeObservable(expected);
  });
});
