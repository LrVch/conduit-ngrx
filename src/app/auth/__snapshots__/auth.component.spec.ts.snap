// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthComponent should compile and show "Have an account?" link 1`] = `
<ng-component
  authErrors$={[Function Store]}
  authType$={[Function Observable]}
  isSubmitting$={[Function Store]}
  route={[Function MockRoute]}
  store={[Function Store]}
  title$={[Function Observable]}
>
  <div
    class="auth-page page"
  >
    <div
      class="container"
    >
      <div
        class="row"
      >
        <div
          class="col-md-6 offset-md-3 col-xs-12"
        >
          <h1>
            Sign up
          </h1>
          
          <p>
            
            <a
              href="/login"
              ng-reflect-link-params="/login"
              ng-reflect-router-link="/login"
            >
              Have an account?
            </a>
            
          </p>
          <app-list-error
            ng-reflect-errors="[object Object]"
          />
          <app-auth-from
            ng-reflect-auth-type="register"
            ng-reflect-is-submitting="false"
          >
            <div>
              a
            </div>
          </app-auth-from>
        </div>
      </div>
    </div>
  </div>
</ng-component>
`;

exports[`AuthComponent should compile and show "Need an account?" link 1`] = `
<ng-component
  authErrors$={[Function Store]}
  authType$={[Function Observable]}
  isSubmitting$={[Function Store]}
  route={[Function MockRoute]}
  store={[Function Store]}
  title$={[Function Observable]}
>
  <div
    class="auth-page page"
  >
    <div
      class="container"
    >
      <div
        class="row"
      >
        <div
          class="col-md-6 offset-md-3 col-xs-12"
        >
          <h1>
            Sign in
          </h1>
          
          <p>
            
            
            <a
              href="/register"
              ng-reflect-link-params="/register"
              ng-reflect-router-link="/register"
            >
              Need an account?
            </a>
          </p>
          <app-list-error
            ng-reflect-errors="[object Object]"
          />
          <app-auth-from
            ng-reflect-auth-type="login"
            ng-reflect-is-submitting="false"
          >
            <div>
              a
            </div>
          </app-auth-from>
        </div>
      </div>
    </div>
  </div>
</ng-component>
`;
